`unnecessary-list-index-lookup` false-positive when iterable is modified in loop

### Bug description

With some code like this:

```python
# pylint: disable=missing-module-docstring

parts = ["a", "b", "c", "d"]
for i, part in enumerate(parts):
    if i == 3:  # more complex condition actually
        parts.insert(i, "X")
    print(part, parts[i])
```

([real-world example](https://github.com/qutebrowser/qutebrowser/blob/1996c4d7e72bb3764cd4ea316f8cf84d4fc32dca/qutebrowser/completion/completer.py#L154-L165)), pylint claims that `parts[i]` could be replaced by `part` - but as running the file shows, that isn't actually the case:

```
a a
b b
c c
d X
d d
```

You might think that this is a bit unorthodox (and I might agree, should probably rewrite this...), but if I'm reading the [Python docs](https://docs.python.org/3.7/reference/compound_stmts.html#the-for-statement) right, the behavior is well-defined (emphasis mine):

> There is a subtlety when the sequence is being modified by the loop (this can only occur for mutable sequences, e.g. lists). An internal counter is used to keep track of which item is used next, and this is incremented on each iteration. When this counter has reached the length of the sequence the loop terminates. This means that if the suite deletes the current (or a previous) item from the sequence, the next item will be skipped (since it gets the index of the current item which has already been treated). Likewise, **if the suite inserts an item in the sequence before the current item, the current item will be treated again the next time through the loop**.

cc @timmartin who added the checker in #5834.

### Configuration

_No response_

### Command used

```shell
pylint x.py
```


### Pylint output

```shell
************* Module x
x.py:7:16: R1736: Unnecessary list index lookup, use 'part' instead (unnecessary-list-index-lookup)
```


### Expected behavior

No error

### Pylint version

```shell
pylint 2.14.1
astroid 2.11.5
Python 3.10.5 (main, Jun  6 2022, 18:49:26) [GCC 12.1.0]
```


### OS / Environment

Archlinux

### Additional dependencies

_No response_

--- a/doc/whatsnew/fragments/6896.false_positive
+++ b/doc/whatsnew/fragments/6896.false_positive
@@ -1,0 +1,3 @@
+``unnecessary-list-index-lookup`` is now more conservative to avoid potential false positives.
+
+Closes #6896

--- a/pylint/checkers/refactoring/refactoring_checker.py
+++ b/pylint/checkers/refactoring/refactoring_checker.py
@@ -2127,7 +2127,7 @@ class RefactoringChecker(checkers.BaseTokenChecker):
         )
         has_nested_loops = next(nested_loops, None) is not None
 
-        # Check if there are any if statemetns within the loop in question;
+        # Check if there are any if statements within the loop in question;
         # If so, we will be more conservative about reporting errors as we
         # can't yet do proper control flow analysis to be sure when
         # reassignment will affect us

--- a/pylint/checkers/refactoring/refactoring_checker.py
+++ b/pylint/checkers/refactoring/refactoring_checker.py
@@ -2127,6 +2127,15 @@ class RefactoringChecker(checkers.BaseTokenChecker):
         )
         has_nested_loops = next(nested_loops, None) is not None
 
+        # Check if there are any if statemetns within the loop in question;
+        # If so, we will be more conservative about reporting errors as we
+        # can't yet do proper control flow analysis to be sure when
+        # reassignment will affect us
+        if_statements = itertools.chain.from_iterable(
+            child.nodes_of_class(nodes.If) for child in children
+        )
+        has_if_statements = next(if_statements, None) is not None
+
         for child in children:
             for subscript in child.nodes_of_class(nodes.Subscript):
                 if isinstance(node, nodes.For) and _is_part_of_assignment_target(
@@ -2170,6 +2179,8 @@ class RefactoringChecker(checkers.BaseTokenChecker):
                         # loops we don't want to report this unless we get to the
                         # end of the loop without updating the collection
                         bad_nodes.append(subscript)
+                    elif has_if_statements:
+                        continue
                     else:
                         self.add_message(
                             "unnecessary-list-index-lookup",

--- a/tests/functional/u/unnecessary/unnecessary_list_index_lookup.py
+++ b/tests/functional/u/unnecessary/unnecessary_list_index_lookup.py
@@ -74,3 +74,10 @@ for idx, val in enumerate(updated_list):
     print(updated_list[idx]) # [unnecessary-list-index-lookup]
     updated_list[idx] -= 1
     print(updated_list[idx])
+
+# Regression test for https://github.com/PyCQA/pylint/issues/6896
+parts = ["a", "b", "c", "d"]
+for i, part in enumerate(parts):
+    if i == 3:  # more complex condition actually
+        parts.insert(i, "X")
+    print(part, parts[i])

